core/blockchain.go

```diff
@@ -2039,21 +2079,67 @@ func (bc *BlockChain) processBlock(parentRoot common.Hash, block *types.Block, s
 		}()
 	}
 
-	// Process block using the parent state as reference point
-	pstart := time.Now()
-	res, err := bc.processor.Process(block, statedb, bc.cfg.VmConfig)
-	if err != nil {
-		bc.reportBlock(block, res, err)
-		return nil, err
+	var res *ProcessResult
+	var resWithMetrics *ProcessResultWithMetrics
+	var ptime, vtime time.Duration
+	if block.Body().AccessList != nil {
+		if block.NumberU64() == 0 {
+			return nil, fmt.Errorf("genesis block cannot have a block access list")
+		}
+		if !validateBAL && !bc.chainConfig.IsGlamsterdam(block.Number(), block.Time()) {
+			bc.reportBlock(block, res, fmt.Errorf("received block containing access list before glamsterdam activated"))
+			return nil, err
+		}
+		// Process block using the parent state as reference point
+		pstart := time.Now()
+		var resCh chan *ProcessResultWithMetrics
+		resCh, err = bc.processor.ProcessWithAccessList(block, statedb, bc.cfg.VmConfig)
```

Comment:
Why does this need to return a channel? Seems like it could be a blocking operation. Also I would probably do the branching for access list vs. not in the `Process` function instead of here since it is really more related to the processing of the block than the broader chain. It should also simplify the changes a bit since both branches of this `if` are quite similar, especially if you then roll the checks from `ValidateProcessResult` into `ValidateState`.